File: langhuan-service/src/main/java/com/langhuan/utils/rag/extractor/DocumentExtractor.java
package com.langhuan.utils.rag.extractor;

import cn.hutool.core.util.IdUtil;
import com.langhuan.model.domain.TFileUrl;
import com.langhuan.service.CacheService;
import com.langhuan.service.MinioService;
import com.langhuan.service.TFileUrlService;
import com.langhuan.utils.other.SecurityUtils;
import jakarta.annotation.PostConstruct;
import jakarta.annotation.Resource;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.apache.poi.xwpf.usermodel.*;
import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.springframework.ai.document.Document;
import org.springframework.ai.reader.tika.TikaDocumentReader;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.core.io.InputStreamResource;
import org.springframework.stereotype.Component;
import org.springframework.web.multipart.MultipartFile;

import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static com.langhuan.common.Constant.CACHE_KEY;
import static com.langhuan.utils.http.GetRequestUtils.sendGetRequest;
import static org.apache.poi.xdgf.util.Util.sanitizeFilename;

import java.io.*;

/**
 * @author Afish
 * @date 2025/7/23 13:42
 */
@Slf4j
@Component
public class DocumentExtractor {

    @Value("${minio.folder.document-img}")
    private String minioFolder; // 在 MinIO 中模拟的“文件夹”前缀

    @Value("${minio.img-bucket-name}")
    private String bucketName;

    @Value("${minio.url}")
    private String minioUrl;

    private String publicBaseUrl; // 用于生成图片访问链接

    @Resource
    private final MinioService minioService;

    @Resource
    private CacheService cacheService;

    @Resource
    private TFileUrlService tFileUrlService;

    public DocumentExtractor(MinioService minioService) {
        this.minioService = minioService;
    }

    @PostConstruct
    public void init() {
        // 确保 bucket 存在
        try {
            minioService.ensureBucketExists(bucketName);
        } catch (Exception e) {
            log.error("Failed to ensure bucket exists", e);
        }

        publicBaseUrl = minioUrl + "/" + bucketName;
        if (!publicBaseUrl.endsWith("/")) {
            publicBaseUrl += "/";
        }
    }

    public String extract(MultipartFile file) throws Exception {
        if (file.getOriginalFilename() != null &&
                file.getOriginalFilename().toLowerCase().endsWith(".docx")) {
            return extractDocxWithImages(file);
        } else {
            return extractWithTika(file);
        }
    }

    private String extractWithTika(MultipartFile file) throws Exception {
        TikaDocumentReader reader = new TikaDocumentReader(new InputStreamResource(file.getInputStream()));
        StringBuilder contentBuilder = new StringBuilder();

        for (Document doc : reader.read()) {
            contentBuilder.append(doc.getFormattedContent()).append("\n");
        }

        String content = contentBuilder.toString();
        if (content.contains("source: Invalid source URI")) {
            content = content.replaceAll(
                    "(?m)^\\s*source: Invalid source URI: InputStream resource.*$", "").trim();
        }
        return content;
    }

    private String extractDocxWithImages(MultipartFile file) throws Exception {
        StringBuilder markdownContent = new StringBuilder();

        // 生成随机 ID 并存入缓存
        int randomId = (int) IdUtil.getSnowflakeNextId();
        cacheService.putId(SecurityUtils.getCurrentUsername() + CACHE_KEY, randomId);
        List<TFileUrl> fileUrlList = new ArrayList<>(); // 用于存储图片信息的列表

        try (InputStream is = file.getInputStream();
                XWPFDocument document = new XWPFDocument(is)) {

            for (IBodyElement element : document.getBodyElements()) {
                if (element instanceof XWPFParagraph) {
                    processParagraph((XWPFParagraph) element, markdownContent, fileUrlList, randomId); // 修改以包含fileUrlList
                    markdownContent.append("\n");
                } else if (element instanceof XWPFTable) {
                    processTable((XWPFTable) element, markdownContent, fileUrlList, randomId);
                    markdownContent.append("\n\n");
                }
            }

            // 批量插入数据库
            if (!fileUrlList.isEmpty()) {
                tFileUrlService.saveBatch(fileUrlList);
            }
        }
        return markdownContent.toString().trim();
    }

    private void processParagraph(XWPFParagraph paragraph, StringBuilder markdownContent, List<TFileUrl> fileUrlList,
            int randomId) {
        for (XWPFRun run : paragraph.getRuns()) {
            String text = run.getText(0);
            if (text != null && !text.isEmpty()) {
                markdownContent.append(text);
            }

            for (XWPFPicture picture : run.getEmbeddedPictures()) {
                handleImage(picture.getPictureData(), fileUrlList, randomId).ifPresent(markdownContent::append);
            }
        }
    }

    private void processTable(XWPFTable table, StringBuilder markdownContent, List<TFileUrl> fileUrlList,
            int randomId) {
        for (XWPFTableRow row : table.getRows()) {
            for (XWPFTableCell cell : row.getTableCells()) {
                for (XWPFParagraph paragraph : cell.getParagraphs()) {
                    processParagraph(paragraph, markdownContent, fileUrlList, randomId);
                }
                markdownContent.append("\t"); // 使用制表符分隔单元格内容
            }
            markdownContent.append("\n"); // 每行结束后换行
        }
    }

    @SneakyThrows
    private Optional<String> handleImage(XWPFPictureData pictureData, List<TFileUrl> fileUrlList, int randomId) {
        String extension = pictureData.suggestFileExtension();
        String fileName = UUID.randomUUID() + "." + extension;
        String objectName = (minioFolder != null ? minioFolder + "/" : "") + fileName;

        try (InputStream inputStream = new ByteArrayInputStream(pictureData.getData())) {
            minioService.handleUpload(objectName, inputStream, pictureData.getData().length, bucketName);
        } catch (Exception e) {
            log.error("Failed to upload image to MinIO", e);
            return Optional.empty();
        }

        // 生成可访问的 URL
        String imageUrl = publicBaseUrl + objectName;

        // 创建 TFileUrl 实体并添加到列表中
        TFileUrl fileUrl = new TFileUrl();
        fileUrl.setFileId(randomId);
        fileUrl.setFUrl(imageUrl);
        fileUrl.setFStatus("临时");
        fileUrlList.add(fileUrl);

        return Optional.of("\n![image](" + imageUrl + ")\n");
    }

    /**
     * 提取 HTML 内容并下载图片到 MinIO
     */
    public String extract(String url) throws Exception {
        String htmlContent = sendGetRequest(url);

        Path dirPath = Paths.get(minioFolder);
        if (!Files.exists(dirPath)) {
            Files.createDirectories(dirPath); // 本地调试用，可选
        }

        org.jsoup.nodes.Document doc = Jsoup.parse(htmlContent);
        Elements imgs = doc.select("img");

        // 生成随机 ID 并存入缓存
        int randomId = (int) IdUtil.getSnowflakeNextId();
        cacheService.putId(SecurityUtils.getCurrentUsername() + CACHE_KEY, randomId);

        // 用于批量插入的列表
        List<TFileUrl> fileUrlList = new ArrayList<>();

        for (Element img : imgs) {
            String src = img.attr("src");
            String alt = img.hasAttr("alt") ? img.attr("alt") : "image";

            if (src.isEmpty())
                continue;

            URL imageUrl = new URL(new URL(url), src);
            String filename = sanitizeFilename(imageUrl.getPath());
            if (!filename.contains(".")) {
                filename += ".jpg"; // 默认扩展名
            }
            String objectName = (minioFolder != null ? minioFolder + "/" : "") + filename;

            try (InputStream in = imageUrl.openStream()) {
                minioService.handleUpload(objectName, in, -1, bucketName); // -1 表示未知大小
            } catch (Exception e) {
                log.error("Failed to upload image from URL to MinIO: {}", imageUrl, e);
                continue;
            }

            // 构建公开访问 URL
            String publicImageUrl = publicBaseUrl + objectName;

            // 创建 TFileUrl 实体
            TFileUrl fileUrl = new TFileUrl();
            fileUrl.setFileId(randomId);
            fileUrl.setFUrl(publicImageUrl);
            fileUrl.setFStatus("临时"); // 或 "temp"，根据你的业务定义
            fileUrlList.add(fileUrl);

            String markdownLink = String.format("![%s](%s%s)", alt, publicBaseUrl, objectName);
            img.replaceWith(new org.jsoup.nodes.TextNode(markdownLink));
        }

        // 批量插入数据库
        if (!fileUrlList.isEmpty()) {
            tFileUrlService.saveBatch(fileUrlList);
        }

        return doc.body().text();
    }
}

File: langhuan-server/src/main/kotlin/com/langhuan/utils/rag/extractor/DocumentExtractor.kt
package com.langhuan.utils.rag.extractor

import cn.hutool.core.util.IdUtil
import com.langhuan.common.Constant.CACHE_KEY
import com.langhuan.model.domain.TFileUrl
import com.langhuan.service.CacheService
import com.langhuan.service.MinioService
import com.langhuan.service.TFileUrlService
import com.langhuan.utils.other.SecurityUtils
import jakarta.annotation.PostConstruct
import jakarta.annotation.Resource
import lombok.SneakyThrows
import org.apache.poi.xwpf.usermodel.*
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import org.jsoup.select.Elements
import org.springframework.ai.document.Document
import org.springframework.ai.reader.tika.TikaDocumentReader
import org.springframework.beans.factory.annotation.Value
import org.springframework.core.io.InputStreamResource
import org.springframework.stereotype.Component
import org.springframework.web.multipart.MultipartFile
import org.slf4j.LoggerFactory
import java.io.BufferedReader
import java.io.ByteArrayInputStream
import java.io.InputStream
import java.io.InputStreamReader
import java.net.URL
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.util.*
import java.net.HttpURLConnection
import java.net.URI


/**
 * @author Afish
 * @date 2025/7/23 13:42
 */
@Component
class DocumentExtractor(
    private val minioService: MinioService
) {

    companion object {
        private val log = LoggerFactory.getLogger(DocumentExtractor::class.java)
    }

    @Value("\${minio.folder.document-img}")
    private lateinit var minioFolder: String // 在 MinIO 中模拟的"文件夹"前缀

    @Value("\${minio.img-bucket-name}")
    private lateinit var bucketName: String

    @Value("\${minio.url}")
    private lateinit var minioUrl: String

    private lateinit var publicBaseUrl: String // 用于生成图片访问链接

    @Resource
    private lateinit var cacheService: CacheService

    @Resource
    private lateinit var tFileUrlService: TFileUrlService

    @PostConstruct
    fun init() {
        // 确保 bucket 存在
        try {
            minioService.ensureBucketExists(bucketName)
        } catch (e: Exception) {
            log.error("Failed to ensure bucket exists", e)
        }

        publicBaseUrl = minioUrl + "/" + bucketName
        if (!publicBaseUrl.endsWith("/")) {
            publicBaseUrl += "/"
        }
    }

    @Throws(Exception::class)
    fun extract(file: MultipartFile): String {
        if (file.originalFilename != null &&
                file.originalFilename!!.lowercase().endsWith(".docx")) {
            return extractDocxWithImages(file)
        } else {
            return extractWithTika(file)
        }
    }

    @Throws(Exception::class)
    private fun extractWithTika(file: MultipartFile): String {
        val reader = TikaDocumentReader(InputStreamResource(file.inputStream))
        val contentBuilder = StringBuilder()

        for (doc in reader.read()) {
            contentBuilder.append(doc.formattedContent).append("\n")
        }

        var content = contentBuilder.toString()
        if (content.contains("source: Invalid source URI")) {
            content = content.replace(
                    "(?m)^\\s*source: Invalid source URI: InputStream resource.*$", "").trim()
        }
        return content
    }

    @Throws(Exception::class)
    private fun extractDocxWithImages(file: MultipartFile): String {
        val markdownContent = StringBuilder()

        // 生成随机 ID 并存入缓存
        val randomId = IdUtil.getSnowflakeNextId().toInt()
        cacheService.putId(SecurityUtils.getCurrentUsername() + CACHE_KEY, randomId)
        val fileUrlList = mutableListOf<TFileUrl>() // 用于存储图片信息的列表

        file.inputStream.use { inputStream ->
            XWPFDocument(inputStream).use { document ->
                val bodyElements = document.bodyElements
                for (element in bodyElements) {
                    if (element is XWPFParagraph) {
                        processParagraph(element, markdownContent, fileUrlList, randomId) // 修改以包含fileUrlList
                        markdownContent.append("\n")
                    } else if (element is XWPFTable) {
                        processTable(element, markdownContent, fileUrlList, randomId)
                        markdownContent.append("\n\n")
                    }
                }

                // 批量插入数据库
                if (fileUrlList.isNotEmpty()) {
                    tFileUrlService.saveBatch(fileUrlList)
                }
            }
        }
        return markdownContent.toString().trim()
    }

    private fun processParagraph(paragraph: XWPFParagraph, markdownContent: StringBuilder, 
                                fileUrlList: MutableList<TFileUrl>, randomId: Int) {
        for (run in paragraph.runs) {
            val text = run.getText(0)
            if (!text.isNullOrEmpty()) {
                markdownContent.append(text)
            }

            for (picture in run.embeddedPictures) {
                handleImage(picture.pictureData, fileUrlList, randomId).ifPresent { markdownContent.append(it) }
            }
        }
    }

    private fun processTable(table: XWPFTable, markdownContent: StringBuilder, 
                            fileUrlList: MutableList<TFileUrl>, randomId: Int) {
        for (row in table.rows) {
            for (cell in row.tableCells) {
                for (paragraph in cell.paragraphs) {
                    processParagraph(paragraph, markdownContent, fileUrlList, randomId)
                }
                markdownContent.append("\t") // 使用制表符分隔单元格内容
            }
            markdownContent.append("\n") // 每行结束后换行
        }
    }

    @SneakyThrows
    private fun handleImage(pictureData: XWPFPictureData, fileUrlList: MutableList<TFileUrl>, randomId: Int): Optional<String> {
        val extension = pictureData.suggestFileExtension()
        val fileName = UUID.randomUUID().toString() + "." + extension
        val objectName = (minioFolder + "/") + fileName

        try {
            ByteArrayInputStream(pictureData.data).use { inputStream ->
                minioService.handleUpload(objectName, inputStream, pictureData.data.size.toLong(), bucketName)
            }
        } catch (e: Exception) {
            log.error("Failed to upload image to MinIO", e)
            return Optional.empty()
        }

        // 生成可访问的 URL
        val imageUrl = publicBaseUrl + objectName

        // 创建 TFileUrl 实体并添加到列表中
        val fileUrl = TFileUrl()
        fileUrl.fileId = randomId
        fileUrl.fUrl = imageUrl
        fileUrl.fStatus = "临时"
        fileUrlList.add(fileUrl)

        return Optional.of("\n![image]($imageUrl)\n")
    }

    /**
     * 发送 GET 请求获取网页内容
     */
    @Throws(Exception::class)
    private fun sendGetRequest(url: String): String {
        val connection = URL(url).openConnection() as HttpURLConnection
        connection.requestMethod = "GET"
        connection.connectTimeout = 10000
        connection.readTimeout = 10000
        connection.setRequestProperty("User-Agent", "Mozilla/5.0")
        
        return try {
            connection.inputStream.use { inputStream ->
                BufferedReader(InputStreamReader(inputStream)).use { reader ->
                    val response = StringBuilder()
                    var line: String?
                    while (reader.readLine().also { line = it } != null) {
                        response.append(line).append("\n")
                    }
                    response.toString()
                }
            }
        } finally {
            connection.disconnect()
        }
    }

    /**
     * 清理文件名，移除不安全字符
     */
    private fun sanitizeFilename(filename: String): String {
        return filename.replace("[^a-zA-Z0-9\\.\\-_]".toRegex(), "_")
    }

    /**
     * 提取 HTML 内容并下载图片到 MinIO
     */
    @Throws(Exception::class)
    fun extract(url: String): String {
        val htmlContent = sendGetRequest(url)

        val dirPath = Paths.get(minioFolder)
        if (!Files.exists(dirPath)) {
            Files.createDirectories(dirPath) // 本地调试用，可选
        }

        val doc = Jsoup.parse(htmlContent)
        val imgElements = doc.select("img")

        // 生成随机 ID 并存入缓存
        val randomId = IdUtil.getSnowflakeNextId().toInt()
        cacheService.putId(SecurityUtils.getCurrentUsername() + CACHE_KEY, randomId)

        // 用于批量插入的列表
        val fileUrlList = mutableListOf<TFileUrl>()

        for (img in imgElements) {
            val src = img.attr("src")
            val alt = if (img.hasAttr("alt")) img.attr("alt") else "image"

            if (src.isEmpty()) continue

            val imageUrl = URL(URL(url), src)
            var filename = sanitizeFilename(imageUrl.path)
            if (!filename.contains(".")) {
                filename += ".jpg" // 默认扩展名
            }
            val objectName = (minioFolder + "/") + filename

            try {
                imageUrl.openStream().use { `in` ->
                    minioService.handleUpload(objectName, `in`, -1, bucketName) // -1 表示未知大小
                }
            } catch (e: Exception) {
                log.error("Failed to upload image from URL to MinIO: {}", imageUrl, e)
                continue
            }

            // 构建公开访问 URL
            val publicImageUrl = publicBaseUrl + objectName

            // 创建 TFileUrl 实体
            val fileUrl = TFileUrl()
            fileUrl.fileId = randomId
            fileUrl.fUrl = publicImageUrl
            fileUrl.fStatus = "临时" // 或 "temp"，根据你的业务定义
            fileUrlList.add(fileUrl)

            val markdownLink = String.format("![%s](%s%s)", alt, publicBaseUrl, objectName)
            img.replaceWith(org.jsoup.nodes.TextNode(markdownLink))
        }

        // 批量插入数据库
        if (fileUrlList.isNotEmpty()) {
            tFileUrlService.saveBatch(fileUrlList)
        }

        return doc.body().text()
    }
}
